parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala

```diff
@@ -106,47 +110,40 @@ object Tokens extends CrfTokenable {
     */
   def postTokenizeTreatment(tokens: Seq[CrfTokenResult]): Map[String, String] = {
-    val groupedTokens = tokens.groupBy(_.label)
+    val groupedTokens = tokens.groupBy(_.label).map { case (label, crfTokenResults) =>
+      label -> crfTokenResults.map(_.value).mkString(" ")
+    }
+
     val postcodeTreatedTokens = postTokenizeTreatmentPostCode(groupedTokens)
     val buildingNumberTreatedTokens = postTokenizeTreatmentBuildingNumber(postcodeTreatedTokens)
-    val buildingNameTreatedTokens = postTokenizeTreatmentBuildingName(buildingNumberTreatedTokens)
-    buildingNameTreatedTokens.map {
-      case (token, values) => (token, values.map(_.value).mkString(" "))
-    }
+    postTokenizeTreatmentBuildingNames(buildingNumberTreatedTokens)
+
   }
 
   /**
     * Normalizes postcode address
     * @param tokens tokens grouped by label
     * @return Map with tokens that will contain normalized postcode address
     */
-  def postTokenizeTreatmentPostCode(tokens: Map[String, Seq[CrfTokenResult]]): Map[String, Seq[CrfTokenResult]] = {
+  def postTokenizeTreatmentPostCode(tokens: Map[String, String]): Map[String, String] = {
 
-    val postcode: Option[String] = tokens.get(Tokens.postcode).map(postcodeTokens => postcodeTokens.map(_.value).mkString(""))
+    // Before analyzing the postcode, we also remove whitespaces so that they don't influence the outcome
+    val postcodeToken: Option[String] = tokens.get(postcode).map(_.replaceAll("\\s", ""))
 
-    postcode match {
-      case Some(concatenatedPostcode) if concatenatedPostcode.length >= 4 =>
-        val postcodeIn = concatenatedPostcode.substring(concatenatedPostcode.length - 3, concatenatedPostcode.length)
-        val postcodeOut = concatenatedPostcode.substring(0, concatenatedPostcode.indexOf(postcodeIn))
+    postcodeToken match {
+      case Some(concatenatedPostcode) if concatenatedPostcode.length >= 5 =>
+        val postcodeInToken = concatenatedPostcode.substring(concatenatedPostcode.length - 3, concatenatedPostcode.length)
+        val postcodeOutToken = concatenatedPostcode.substring(0, concatenatedPostcode.indexOf(postcodeInToken))
 
-        val tokensWithPostcodeUpdated = tokens.updated(Tokens.postcode, Seq(
-          CrfTokenResult(
-            value = s"$postcodeOut $postcodeIn",
-            label = Tokens.postcode
-          )
-        ))
+        val tokensWithPostcodeUpdated = tokens.updated(postcode, s"$postcodeOutToken $postcodeInToken")
 
         tokensWithPostcodeUpdated ++ Map(
-          Tokens.postcodeOut -> Seq(CrfTokenResult(
-            value = postcodeOut,
-            label = Tokens.postcodeOut
-          )),
-          Tokens.postcodeIn -> Seq(CrfTokenResult(
-            value = postcodeIn,
-            label = Tokens.postcodeIn
-          ))
+          postcodeOut -> postcodeOutToken,
+          postcodeIn -> postcodeInToken
         )
 
+      case Some(concatenatedPostcode) => tokens + (postcodeIn -> concatenatedPostcode)
```

Comment:
that should have postcodeOut instead of postcodeIn? The first component e.g. RH1, or PO15, or whatever. Not of the format \d[A-Z]{2}.